metadata-validator.py

```diff
@@ -0,0 +1,227 @@
+    return not [logError(element, COMPILATION_ERROR_MESSAGE, element[NAME_KEY]) \
+                for src in glob.glob(os.path.join(path, "*.c")) if compileFile(src, path) != 0]
+
+
+def checkLibraryCreationValid(element, path):
+    ar_command = AR + " " + AR_ARGS + " " + os.path.join(path, "lib.a")  +  " ".join(glob.glob(os.path.join(path + "*.o")))
+    #print(ar_command)
+
+    if os.system(ar_command) != 0:
+        logError(element, CODE_GENERATION_ERROR_MESSAGE, element[NAME_KEY])
+        return False
+    else:
+        return True
+
+
+def checkElementCompilationValid(element, element_lib):
+    dependencies = set(getRecursiveDependency(element, element_lib))
+    path = os.path.join(BUILD_PATH, makeValidFilename(element[NAME_KEY]))
+    print (path)
+    os.makedirs(path)
+    if not dependencies: return True
+    if not checkCCodeGenerationValid(element, dependencies, path): return False
+    if not checkCompilationValid(element, path): return False
+    if not checkLibraryCreationValid(element, path): return False
+    return True
+
+def checkCodeValid(element_lib):
+    return all(checkElementCompilationValid(element, element_lib) for element in element_lib)
+
+
+def loadLibrary(path):
+    metadata = []
+    files = glob.glob(os.path.join(path, "*/" + METADATA_FILENAME))
+    for filepath in files:
+        with open(filepath) as f:
+            metadata.append({METADATA_KEY: json.load(f), METADATA_PATH_KEY: filepath.replace(METADATA_FILENAME, "")})
+    
+    element_lib = []
+    for meta in metadata:
+        for submodule in meta[METADATA_KEY].get(SUBMODULES_KEY, []):
+            for element in submodule.get(ELEMENTS_KEY, []):
+                element[METADATA_PATH_KEY] = meta[METADATA_PATH_KEY]
+                element[MODULE_NAME_KEY] = meta[METADATA_KEY][NAME_KEY]
+                element_lib.append(element)
+
+    return element_lib
+
+
+def validateMetadata():
+    try:
+        if(os.path.isdir(BUILD_PATH)):
+            shutil.rmtree(BUILD_PATH)
+
+        os.makedirs(BUILD_PATH)
+    except OSError:
+        print("Error when creating directory: %s" % (BUILD_PATH))
+        return ERROR_CODE
+    
+    element_lib = loadLibrary(LIBRARY_PATH)
+
+    print("\nVerifying internal metadata dependencies")
+    if checkInternalDependenciesValid(element_lib):
+        print(VALIDATION_PASSED_MESSAGE)
+    else:
+        print(VALIDATION_FAILED_MESSAGE)
+        return ERROR_CODE
+
+    print("\nVeryfing external metadata dependencies")
+    if checkExternalDependenciesValid(element_lib):
+        print(VALIDATION_PASSED_MESSAGE)
+    else:
+        print(VALIDATION_FAILED_MESSAGE)
+        return ERROR_CODE
+
+    print("\nGenerating and building C files")
+    if checkCodeValid(element_lib):
+        print(VALIDATION_PASSED_MESSAGE)
+    else:
+        print(VALIDATION_FAILED_MESSAGE)
+        return ERROR_CODE
+
+    return SUCCESS_CODE
+
+if __name__ == "__main__":
+    sys.exit(validateMetadata())
```

Comment:
could be useful to have lib path passed as program argument, but mkay

Reply:
Added both args to validateMetadata(). Default is directory with script so I didn't touch this line